### PR TITLE
Finish SdlPsm

### PR DESCRIPTION
**Pull request by joeygrover**
This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Summary
- Finished the `SdlPsm` class
- Started to integrate `SdlPsm` into other classes
- Some work needs to be looked over before merging to mainline, but this PR only merges to feature branch

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform) and this pull request adheres to the [Contributing Guide](./CONTRIBUTING.md).

### CODE REVIEW

File: lib/node/transport/WebsocketServer.js [hunk 1]
```diff
@@ -56,7 +57,29 @@ class WebsocketServer extends TransportBase {
...
                 console.log('is a')
+                stateProgress = false;
+                for(i = 0; i < message.length; i++){
+                    stateProgress = this._sdlPsm.handleByte(message[i]); //TODO make sure it can be accessed like this
+
+                    if(!stateProgress){ //We are trying to weed through the bad packet info until we get something
+                        this._sdlPsm.reset();
+                    }else if( this._sdlPsm.getState() === SdlPsm.FINISHED_STATE){
+                        packet = this._sdlPsm.getFormedPacket();
+                        if(this._transportListener != null && packet != null){
```
Comment (nickschwab) [inline lib/node/transport/WebsocketServer.js]:
```suggestion
                        if (this._transportListener != null && packet != null) {
```

File: lib/node/transport/WebsocketServer.js [hunk 2]
```diff
@@ -56,7 +57,29 @@ class WebsocketServer extends TransportBase {
...
+                            this._transportListener.onPacketReceived(packet);
+                        }
+
+                        this._sdlPsm.reset();
+                    }
+                }
             });
+
+            //Back to 'connection'
+            if(this._transportListener != null ){
```
Comment (nickschwab) [inline lib/node/transport/WebsocketServer.js]:
```suggestion
            if (this._transportListener != null ) {
```

File: lib/js/transport/SdlPsm.js [hunk 3]
```diff
@@ -31,231 +31,11 @@
...
-
-
-
-
-
-
-
-
-class SdlPsm {
+class SdlPsm {							//TODO check to see if it's better to use '==' or '==='
```
[the remark this answers is not in the document]
Comment (joeygrover) [reply lib/js/transport/SdlPsm.js]: I think leaving this note for now is fine, as I want to be sure we understand what data types we will be receiving based on different platforms transports (node.js vs vanilla js).

File: lib/js/transport/SdlPsm.js [hunk 4]
```diff
@@ -344,196 +124,152 @@ class SdlPsm {
...
             case SdlPsm.SERVICE_TYPE_STATE:
-                // TODO
+
+				this._serviceType = rawByte & 0xFF;
+				return SdlPsm.CONTROL_FRAME_INFO_STATE;
 
             case SdlPsm.CONTROL_FRAME_INFO_STATE:
-                // TODO
+				
+				this._controlFrameInfo = rawByte & 0xFF;
```
Comment (joeygrover) [inline lib/js/transport/SdlPsm.js]:
```suggestion
		this._controlFrameInfo = rawByte & 0xFF;
```

File: lib/js/transport/SdlPsm.js [hunk 5]
```diff
@@ -344,196 +124,152 @@ class SdlPsm {
...
+				}
 
-            case SdlPsm.MESSAGE_1_STATE:
-                // TODO
+				if(this._version == 1){ //Version 1 packets will not have message id's
+					if(this._dataLength == 0){
+						return SdlPsm.FINISHED_STATE; //We are done if we don't have any payload
+					}
+					if(this._dataLength <= SdlProtocolBase.V1_V2_MTU_SIZE - SdlProtocolBase.V1_HEADER_SIZE){ // sizes from protocol/SdlProtocol.java
+						this._payload = new byte[this._dataLength];
```
Comment (joeygrover) [inline lib/js/transport/SdlPsm.js]:
```suggestion
						this._payload = new Uint8Array(this._dataLength);
```

File: lib/js/transport/SdlPsm.js [hunk 6]
```diff
@@ -344,196 +124,152 @@ class SdlPsm {
...
 
             case SdlPsm.MESSAGE_4_STATE:
-                // TODO
+				
+				this._messageId += ( rawByte& 0xFF );
+				if(this._dataLength == 0){
+					return SdlPsm.FINISHED_STATE; //We are done if we don't have any payload
+				}
+				try{
+					this._payload = new byte[this._dataLength];
```
Comment (joeygrover) [inline lib/js/transport/SdlPsm.js]:
```suggestion
					this._payload = new Uint8Array(this._dataLength);
```

File: lib/js/transport/SdlPsm.js [hunk 7]
```diff
@@ -344,196 +124,152 @@ class SdlPsm {
...
 
             case SdlPsm.DATA_PUMP_STATE:
-                // TODO
+				
+				this._payload[this._dataLength - this._dumpSize] = rawByte;
+				this._dumpSize--;
+				//Do we have any more bytes to read in?
+				if(this._dumpSize > 0){
+					return SdlPsm.DATA_PUMP_STATE;
+				}
```
[the remark this answers is not in the document]
Comment (joeygrover) [reply lib/js/transport/SdlPsm.js]: I don't think we want to get rid of that curly brace, right? Just the extra indentation?

File: lib/js/transport/TransportListener.js [hunk 8]
```diff
@@ -77,6 +79,26 @@ class TransportListener {
     getOnPacketReceived() {
         return this._onPacketReceived;
     }
+
+    //TODO check if this works, if so we likely don't need the getters
```
[the remark this answers is not in the document]
Comment (joeygrover) [reply lib/js/transport/TransportListener.js]: Ok, so we should add null checks to each call? Is this something done in javascript though or is returning the function itself more commonly done? I think this method is less verbose, but I'd rather stick to the platform norms.

[the remark this answers is not in the document]
Comment (nickschwab) [reply lib/js/transport/TransportListener.js]: Executing the function is ideal here, so I recommend we either wrap this with a type check for a silent failure (example below), or throw a developer-friendly error if the type check fails.
`if (typeof this._onTransportConnected === 'function') this._onTransportConnected();`

**Pull request merged**